### ISSUE TITLE
Email validation

### ISSUE THREAD

**Issue by skylord123** [posativ/isso#562]
It would be cool to add email validation.

It would help with spam and also the possibility of someone using someone else's email.

**Comment by ix5** [comment]
Will the site admin receive a notification _before_ or _after_ the commenter has verified their email?

Things to think about in terms of the backend:
- `mode` in database
- Can site admins manually override not-yet-verified commenters?
- Where do not-yet-verified comments show up in the admin?

Other aspects:
- How long will verification be stored? At the moment, cookies have a lifespan of 15 minutes (`max-age`). Do we want to ask commenters to re-verify their emails after that cookie has expired in order for them to add a new comment?
- Will the "token" they receive in their email be valid beyond the 15 minutes, allowing them to re-obtain a cookie that lets them edit _any_ or just _recent_ comments with their email?

**Comment by p1slave** [comment]
@jelmer I want to fix this one too. I lost count of how many times people enter the wrong email such as `gmail,com` causing the confirmation email to bounce back. 

My idea is to ask user to verify their email before they can leave a comment. Once the email address verified, it's marked in the database as verified email and they don't have to verify it again next time.

**Comment by jelmer** [comment]
Supporting this makes sense I think.

> Will the site admin receive a notification before or after the commenter has verified their email?

I think this should be after the commenter has verified their email, otherwise spammers without valid email addresses can still bother admins. Whether verification is necessary should be a per-site boolean option.

> How long will verification be stored? At the moment, cookies have a lifespan of 15 minutes (max-age). Do we want to ask commenters to re-verify their emails after that cookie has expired in order for them to add a new comment?

This makes sense to me - the 15 minutes are intentional, and making that e.g. a permanent cookie is a change away from isso's current minimal impact behaviour.

> @jelmer I want to fix this one too. I lost count of how many times people enter the wrong email such as gmail,com causing the confirmation email to bounce back.

FWIW this wouldn't make a difference for when you'd receive bounces, as the emails for confirmation emails can bounce in the exact same way.

> My idea is to ask user to verify their email before they can leave a comment. Once the email address verified, it's marked in the database as verified email and they don't have to verify it again next time.

Wouldn't that just allow people to enter somebody else's email address if that email address has been successfully verified and used to post a comment?

**Comment by p1slave** [comment]
If the user enters a legit email address, there will be no bounces. My point is that we need a way to make sure the user enters the correct email address by double confirmation, otherwise they won't receive email notification when someone replies to their comments.

>Wouldn't that just allow people to enter somebody else's email address if that email address has been successfully verified and used to post a comment?

I thought about that too and it's not a big concern since the email of a user won't be shown in public. If there is no email validation at all, the user can enter any email address they want, then if you approve the comment with an incorrect email address, the notification email could be sent to the wrong person or bounces back.

**Comment by jelmer** [comment]
> If the user enters a legit email address, there will be no bounces. My point is that we need a way to make sure the user enters the correct email address by double confirmation, otherwise they won't receive email notification when someone replies to their comments.

What I mean is that sending out verification emails will also lead to bounces if the email addresses are invalid, so that doesn't really change anything in terms of receiving bounces - if that's the problem you're trying to solve.

> >Wouldn't that just allow people to enter somebody else's email address if that email address has been successfully verified and used to post a comment?
> 
> I thought about that too and it's not a big concern since the email of a user won't be shown in public. If there is no email validation at all, the user can enter any email address they want, then if you approve the comment with an incorrect email address, the notification email could be sent to the wrong person or bounces back.

That IMO defeats one of the purposes of using email validation, what's the point of validating email addresses if you can't know for sure that the person owning the email address has actually posted a comment? 

Also, I would argue that finding a valid email address is trivial. For example, on my blog I've responded to people's comments and my email address is literally mentioned it the about section.

**Comment by p1slave** [comment]
There are always bounces since some users deliberately enter a wrong email address like `123@gmail.com` to test out things. The main purpose of email validation is to make sure the user will receive email notification when I reply to them and also deter bots to some extent.

That's a good point. I guess you have to set an expiration time like 15 minutes and after that the user has to verify it again. 

By the way, is there a way for the user to delete an old comment when all the cookies expire? This feature also requires email validation if we want to have it implemented, right? I'm trying to be GDPR compliant:)

**Comment by jelmer** [comment]
> There are always bounces since some users deliberately enter a wrong email address like `123@gmail.com` to test out things. The main purpose of email validation is to make sure the user will receive email notification when I reply to them and also deter bots to some extent.

FWIW I brought this up because you mentioned bounces as argument for doing email verification. Did I misunderstand what you meant?

> That's a good point. I guess you have to set an expiration time like 15 minutes and after that the user has to verify it again. 

> By the way, is there a way for the user to delete an old comment when all the cookies expire? This feature also requires email validation if we want to have it implemented, right? I'm trying to be GDPR compliant:)

No, they'll have to contact you separately today and then you can delete their comments from the web UI. Providing email is optional either way - and they can lose access to their cookies. 

GDPR requests are likely to cover your entire site either way, not just isso.

**Comment by p1slave** [comment]
Sorry for the misunderstanding. The main point is that they enter the wrong email address and won't receive email notification. Email bounce is one of the consequences of entering wrong email and it is not a big deal.

So we still want to implement the email validation feature and add an expiration time to it?

**Comment by jelmer** [comment]
Yeah, implementing email validation seems useful to me. Id keep optional though, and perhaps simple for now: require confirming each comment by email by making the user click on a link in the email.
